rfcs/rfcOVID-0019.md

```diff
@@ -0,0 +1,329 @@
+method:
+
+```perl
+package My::Module {
+    use v5.38;
+
+    sub import ($class, @args) {
+        ...
+    }
+    sub bar  :export           {...} # @EXPORT_OK
+    sub baz  :export(strings)  {...} # %EXPORT_TAGS
+    sub bay  :export(strings)  {...} # %EXPORT_TAGS
+    sub quux :export(DEFAULT)  {...} # @EXPORT
+    sub whee                   {...} # not exported
+}
+```
+
+If an `import()` method is defined in the package, it will need to call
+`UNIVERSAL::import()` to have the lexical importing handled correctly.
+
+```perl
+sub import ($class, @args) {
+    my ( $import, $mine ) = separate_args(@args);
+    # do stuff with $mine
+    UNIVERSAL::import($import->@*);
+}
+```
+
+For the above, how do we tell `UNIVERSAL::import` which class its importing
+into? I suppose we could do `local @_ = $import->@*; goto &UNIVERSAL::import`,
+but that's finicky, it's going to break, and we're trying to discourage use of
+`@_` in signatured subs. I expect there's some new work done in Perl which I'm
+missing here.
+
+## Security Implications
+
+None anticipated, but by having the imports being lexical, it means outside
+code might have a harder time replacing these functions. Thus, it might be
+more secure.
+
+## Examples
+
+```perl
+package My::Module {
+    use v5.38;
+
+    sub bar  :export           {...} # @EXPORT_OK
+    sub baz  :export(strings)  {...} # %EXPORT_TAGS
+    sub bay  :export(strings)  {...} # %EXPORT_TAGS
+    sub quux :export(DEFAULT)  {...} # @EXPORT
+    sub whee                   {...} # not exported
+}
+
+use My::Package 'bar';     # imports bar() and quux
+
+use My::Package ':strings' # imports baz, bay, and quux
+
+use My::Package;           # imports quux
+
+use My::Package ();        # imports nothing
+
+use My::Package qw/bar :strings -quux/; # don't import quux
+
+use My::Module 'whee'; # fatal error
+```
+
+## Prototype Implementation
+
+None that I'm aware of, though
+[Perl6::Export::Attrs](https://metacpan.org/pod/Perl6::Export::Attrs) seems
+the closest.
+
+## Future Scope
+
+We might want to allow variables and constants to be exported. We might always
+want export tags to contain other export tags to allow "grouping" of groups.
+
+The 'is not exported' error message could list what is actually exported by
+the module.
+
+An introspection API for procedural code (similar to a MOP), could be
+provided. As part of this, it could provide an export spec. This could allow
+Perl to start providing "smart" documentation that reads the code in addition
+to POD.
+
+We could also determine a syntax as importing as a different name. This should
+be safe because they're lexical imports.
+
+## Rejected Ideas
+
+This was part of larger "AMORES" pre-RFC to improve the overall state of
+procedural code. That pre-RFC was rejected, but I was asked to put this into
+its own RFC.
+
+## Open Issues
+
+None?
+
+## Copyright
+
+Copyright (C) 20222-23, Curtis "Ovid" Poe
```

Comment:
Hopefully we don't have to wait quite that long.

Reply:
Fixed (oops).